### PR TITLE
Add support for mtk mt8186 platform

### PR DESCRIPTION
**Pull request by tinghan-shen**
Hi,

I'm a chromebook project developer at Mediatek and working on SOF FW on mt8186.

The other co-worker is Allen Cheng (KuanHsun Cheng).

There are some drivers and platform codes we have been verified on mt8186 EVB.

We want to add mt8186 platform support in sof project.

mt8186: Cadence HiFi-5 DSP, single core, 4 dais , 4 dmas and 48k 16 bits.

Similar to mt8195 except changing the DSP core version.

Please let me know if there is any questions. 

Thanks.

### CODE REVIEW

**Comment by sofci** [comment]
Can one of the admins verify this patch?
> reply **test&nbsp;this&nbsp;please** to run this test once

**Comment by gkbldcig** [comment]
Can one of the admins verify this patch?

**Comment by afq984** [comment]
test this please

**Comment by afq984** [comment]
approved the github actions, not sure if I can run intel CI though

**Review by dbaluta — APPROVED** [review]

**Comment by dbaluta** [comment]
Looks good to me. Do you also have a gcc based toolchain for this? It would be useful to add it in the CI build scripts so that we can catch early if something breaks.

**Comment by tinghan-shen** [comment]
> Looks good to me. Do you also have a gcc based toolchain for this? It would be useful to add it in the CI build scripts so that we can catch early if something breaks.

Ok, I'll prepare a patch for this. 
thanks the advice.

**Comment by dbaluta** [comment]
> Ok, I'll prepare a patch for this.
> thanks the advice.

You can have a look at https://github.com/thesofproject/crosstool-ng/ for all suppported platforms. I also think we have a repo to upload the toolchain binary?

@lgirdwood i don't remember where we store the toochains.

**Comment by lgirdwood** [comment]
> > Ok, I'll prepare a patch for this.
> > thanks the advice.
> 
> You can have a look at https://github.com/thesofproject/crosstool-ng/ for all suppported platforms. I also think we have a repo to upload the toolchain binary?
> 
> @lgirdwood i don't remember where we store the toochains.

We need to do

1) add overlay to xtensa-overlays repo
2) add config file to crosstool-ng repo (binaries are uploaded to ct-ng downloads section)
3) add binaries to container usually @marc-hb or @greg-intel can do this.
4) add to CI (same folks as 3).

**Comment by lgirdwood** [comment]
> @lgirdwood One thing needs your advise. May I know if qemu is required for CI? or qemu is for git actions? We are preparing qemu for mt8195 now. [wenliangwu/qemu@5f259f1](https://github.com/wenliangwu/qemu/commit/5f259f1ca05338ec584aef57b99ee50c13962c7b) , which is based on v5. 2.0. I am wondering what thing need to do when the qemu patch are ready. Thanks in advance.

This can go in the qemu branch and then be added to the CI container. @marc-hb and @greg-intel can help with this.

Fwiw, we are moving towards Renode for emulation (timeline TBD) as it has several key advantages over qemu for our needs. @cujomalainey  can give you more details.

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
@kkarask can you comment on the CI, this PR should not fail any Intel specific tests.

**Comment by lgirdwood** [comment]
SOFCI TEST

**Comment by cujomalainey** [comment]
> Fwiw, we are moving towards Renode for emulation (timeline TBD) as it has several key advantages over qemu for our needs. @cujomalainey  can give you more details.

Unfortunately the timeline has been disrupted a bit due to budget planning schedules being misaligned. Only basic xtensa is working right now with zephyr. Working to get more added.

**Comment by marc-hb** [comment]
> Do you also have a gcc based toolchain for this?

>  May I know if qemu is required for CI? or qemu is for git actions? 

Please have a look at the discussions and (small) changes in #5202 that was just merged.

File: scripts/xtensa-build-all.sh
Comment (marc-hb) [inline scripts/xtensa-build-all.sh:11]: I don't think `xtensa-mt8186-elf` is available yet, is it? If it is available then please add `mt8186` to `.github/workflows/pull-request.yml`. If not yet then please move `mt8186` to `SUPPORTED_PLATFORMS` for now.

**Review by marc-hb — COMMENTED** [review]
Most of this PR looks like copy/paste of `mt8195`

**Comment by marc-hb** [comment]
@kkarask  https://sof-ci.01.org/sof-pr-viewer/#/build/PR5239/build8393711 has very strange and new failures again... is there some major activity going on?

File: src/platform/mt8186/include/platform/drivers/interrupt.h
Comment (kuanhsuncheng) [inline src/platform/mt8186/include/platform/drivers/interrupt.h:5]: Hi Tinghan,

Please change this to right person. thanks.

File: scripts/xtensa-build-all.sh
Comment (kuanhsuncheng) [reply scripts/xtensa-build-all.sh:11]: Hi marc, I check the current .github/workflows/pull-request.yml. there is no mt8195 strategy. should we update also ?

**Review by kuanhsuncheng — COMMENTED** [review]
I check the current .github/workflows/pull-request.yml. there is no mt8195 strategy. should we update also ?

File: src/drivers/mediatek/mt8186/ipc.c
Comment (lyakh) [inline src/drivers/mediatek/mt8186/ipc.c:158]: no error handling?

File: src/drivers/mediatek/mt8186/timer.c
Comment (lyakh) [inline src/drivers/mediatek/mt8186/timer.c:76]: this is in units, half of what was read in line 61 from `MTK_OSTIMER_CUR_L` because of multiplication in line 65, is this correct?

File: src/drivers/mediatek/mt8186/timer.c
Comment (lyakh) [inline src/drivers/mediatek/mt8186/timer.c:213]: usually the order would be inverted in enable / disable pairs of calls?

File: src/platform/mt8186/lib/clk.c
Comment (lyakh) [inline src/platform/mt8186/lib/clk.c:44]: superfluous internal parentheses in 2 lines above

File: src/platform/mt8186/lib/clk.c
Comment (lyakh) [inline src/platform/mt8186/lib/clk.c:56]: there's no shared code between the two muxes and the caller always uses a constant for the `mux_id`, maybe just make two separate functions?

File: src/drivers/mediatek/mt8186/timer.c
Comment (tinghan-shen) [reply src/drivers/mediatek/mt8186/timer.c:76]: This is intended. The underlying HW is a 32-bit count-down timer.
It's used to set the difference between the current tick and next tick triggering reschedule.

**Comment by marc-hb** [comment]
> I check the current .github/workflows/pull-request.yml. there is no mt8195 strategy. should we update also ?

If you have the gcc toolchain ready in the Docker image then yes of course.

**Comment by lgirdwood** [comment]
@tinghan-shen @kuanhsuncheng please someone from MTK approve PR before merge.

**Comment by cujomalainey** [comment]
@yaochunhung

File: src/platform/mt8186/include/platform/platform.h
Comment (cujomalainey) [inline src/platform/mt8186/include/platform/platform.h:28]: outstanding todo?

**Review by kuanhsuncheng — APPROVED** [review]

**Comment by tinghan-shen** [comment]
Here are the PRs for mt8186 gcc toolchain.

[https://github.com/thesofproject/crosstool-ng/pull/12](https://github.com/thesofproject/crosstool-ng/pull/12)
[https://github.com/thesofproject/crosstool-ng/pull/12](https://github.com/thesofproject/xtensa-overlay/pull/12)

File: src/platform/mt8186/include/platform/drivers/idc.h
Comment (tinghan-shen) [reply src/platform/mt8186/include/platform/drivers/idc.h:17]: This file is added to solve build errors. mt8186 doesn't enable the configuration to build src/idc/idc.c.

File: src/drivers/mediatek/mt8186/interrupt.c
Comment (tinghan-shen) [reply src/drivers/mediatek/mt8186/interrupt.c:134]: It's because some platforms also use it. I thought that maybe it's the way for IRQ handling at here.

for example,
[https://github.com/thesofproject/sof/blob/main/src/drivers/amd/renoir/interrupt.c#L128](https://github.com/thesofproject/sof/blob/main/src/drivers/amd/renoir/interrupt.c#L128)

[https://github.com/thesofproject/sof/blob/main/src/drivers/imx/interrupt-irqsteer.c#L324](https://github.com/thesofproject/sof/blob/main/src/drivers/imx/interrupt-irqsteer.c#L324)

File: scripts/xtensa-build-all.sh
Comment (ranj063) [reply scripts/xtensa-build-all.sh:15]: still not added to the list above

File: src/arch/xtensa/configs/mt8186_defconfig
Comment (tinghan-shen) [reply src/arch/xtensa/configs/mt8186_defconfig:5]: After consult with audio team, MTK doesn't use these functions currently.

File: src/platform/mt8186/include/platform/drivers/interrupt.h
Comment (yaochunhung) [inline src/platform/mt8186/include/platform/drivers/interrupt.h:23]: align

File: src/platform/mt8186/include/platform/lib/memory.h
Comment (yaochunhung) [inline src/platform/mt8186/include/platform/lib/memory.h:15]: remove it since no use

File: src/platform/mt8186/include/platform/lib/memory.h
Comment (yaochunhung) [inline src/platform/mt8186/include/platform/lib/memory.h:19]: remove it

File: src/platform/mt8186/lib/clk.c
Comment (yaochunhung) [inline src/platform/mt8186/lib/clk.c:20]: Need to generate another unique uuid for mt8186, don't use the same one with mt8195.

File: src/drivers/mediatek/mt8186/interrupt.c
Comment (yaochunhung) [inline src/drivers/mediatek/mt8186/interrupt.c:26]: Need to generate another unique uuid for mt8186, don't use the same one with mt8195.

File: src/drivers/mediatek/mt8186/ipc.c
Comment (yaochunhung) [inline src/drivers/mediatek/mt8186/ipc.c:38]: Need to generate another unique uuid for mt8186, don't use the same one with mt8195.

File: src/platform/mt8186/platform.c
Comment (yaochunhung) [inline src/platform/mt8186/platform.c:137]: should be 6 , I think it also can use #define ARRAYSIZE(a) (sizeof(a) / sizeof(*(a))) for it?

**Review by yaochunhung — COMMENTED** [review]
@tinghan-shen @kuanhsuncheng  please see my comments and feel free to discuss with me. thanks.

**Review by kuanhsuncheng — APPROVED** [review]

**Review by yaochunhung — APPROVED** [review]

**Comment by lgirdwood** [comment]
@kkarask can you check CI, as this PR is for MTK platform only and should not fail Intel tests. Thanks !

File: src/lib/notifier.c
Comment (lgirdwood) [inline src/lib/notifier.c:24]: any reason why this has changed ?

File: src/drivers/mediatek/mt8186/interrupt.c
Comment (cujomalainey) [reply src/drivers/mediatek/mt8186/interrupt.c:134]: Interesting @lgirdwood this seems rather inflated, what are your thoughts here?

**Comment by yaochunhung** [comment]
> > Do you also have a gcc based toolchain for this?
> 
> > May I know if qemu is required for CI? or qemu is for git actions?
> 
> Please have a look at the discussions and (small) changes in #5202 that was just merged.

@marc-hb  Thanks for providing information.  @lgirdwood has already helped to merge mt8195 qemu patch https://github.com/thesofproject/qemu/commit/b85c079bbac0a457f7d89dc1ba5dac30151bdf5b. but I am not sure how to add mt8195 to thesofproject/sofqemu (it looks like here https://hub.docker.com/r/thesofproject/sofqemu/tags )and still anything to do except add platform information to pull-request.yml. Besides, may I need to add docker_build like 
https://github.com/thesofproject/sof/commit/66234f02c3bf62ca8042c8db9e815461c7667661#diff-edde3c23b72c58bc38db847aede15fb3e20078318397825da9a7ef2fa48801b5. Please help to advice. Thanks.

File: src/lib/notifier.c
Comment (tinghan-shen) [reply src/lib/notifier.c]: this is a mistake. I'm sorry.

File: src/drivers/mediatek/mt8186/interrupt.c
Comment (lgirdwood) [reply src/drivers/mediatek/mt8186/interrupt.c:134]: Yeah, these could be optimized. Lets merge this first and get the build results as we could make the same change to all platforms. @tinghan-shen pls can you try this change and report any size changes to the binary. Thanks.

**Review by lgirdwood — APPROVED** [review]

File: src/drivers/mediatek/mt8186/interrupt.c
Comment (cujomalainey) [reply src/drivers/mediatek/mt8186/interrupt.c:134]: I think NXP would see the biggest effect since they have multiple IRQs @dbaluta

Comment (tinghan-shen) [reply src/drivers/mediatek/mt8186/interrupt.c:134]: I tried to build mt8186 on both gcc11 and xcc, and compare the size of sof.ri.
No difference on gcc11.
64 bytes larger on xcc after remove "inline".

**Pull request merged**
